include/shared/game/servergamestate.hpp

```diff
@@ -0,0 +1,280 @@
+#pragma once
+
+#include "shared/utilities/typedefs.hpp"
+#include "shared/utilities/config.hpp"
+#include "shared/utilities/smartvector.hpp"
+#include "shared/game/gamelogic/object.hpp"
+
+#include <string>
+#include <vector>
+#include <chrono>
+#include <unordered_map>
+#include <queue>
+
+//	From sharedgamestate.hpp
+struct SharedGameState;
+
+enum class GamePhase {
+	TITLE_SCREEN,
+	LOBBY,
+	GAME
+};
+
+/**
+ * @brief Information about the current lobby of players.
+ */
+struct Lobby {
+	//	TODO: Perhaps instead of a mapping from EntityID -> string, the mapping
+	//	could eventually be EntityID -> Player (where Player derives from
+	//	Object)?
+
+	/**
+	 * @brief A hash table that maps from player's EntityID to their names.
+	 */
+	std::unordered_map<EntityID, std::string> players;
+
+	/**
+	 * @brief The maximum number of players that this game instance can support.
+	 */
+	int max_players;
+
+	//	TODO: Add a player role listing? I.e., which player is playing which
+	//	character and which player is playing as the Dungeon Master?
+};
+
+/*	Constants	*/
+#define	FIRST_TIMESTEP			0
+#define TIMESTEP_LEN			std::chrono::milliseconds(30)
+#define	MAX_PLAYERS				4
+
+//	SmartVector capacities
+#define	MAX_NUM_OBJECTS			100
+#define MAX_NUM_BASE_OBJECTS	100
+
+/**
+ * @brief The ServerGameState class contains all abstract game state data and
+ * logic for a single game state instance (i.e., for one match played by 4
+ * players).
+ */
+class ServerGameState {
+public:
+	/**
+	 * @brief Creates a ServerGameState instance. The intial GamePhase is set to
+	 * Lobby.
+	 */
+	ServerGameState();
+
+	/**
+	 * @brief Creats a ServerGameState instance and sets the initial game phase
+	 * to the given GamePhase.
+	 * @param start_phase GamePhase that the new ServerGameState instance will
+	 * start in.
+	 */
+	ServerGameState(GamePhase start_phase);
+
+	ServerGameState(GamePhase start_phase, GameConfig config);
+
+	~ServerGameState();
+
+	/**
+	 * @brief Updates this ServerGameState from the current timestep to the
+     * next one.
+	 */
+	void update();
+
+	//	TODO: Add specific update methods (E.g., updateMovement() to update
+	//	object movement)
+
+	void updateMovement();
+
+	/*	SharedGameState generation	*/
+
+	//	TODO: Modify this function to dynamically allocate a SharedGameState
+	//	instance and return a pointer (perhaps a unique pointer) to it
+	/**
+	 * @brief Generate a SharedGameState object from this ServerGameState
+	 * instance.
+	 * @return ShareGameState instance that represents this ServerGameState
+	 * instance.
+	 */
+	SharedGameState generateSharedGameState();
```

Comment:
I like this approach of generating the `SharedGameState` from the version the server has.